src/libraries/System.Text.Json/gen/src/JsonSourceGenerator.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json.SourceGeneration.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
+    /// Generates source code to optimize serialization and deserialization with JsonSerializer.
+    /// </summary>
+    [Generator]
+    public sealed class JsonSourceGenerator : ISourceGenerator
+    {
+        private JsonSourceGeneratorHelper? _helper;
+
+        /// <summary>
+        /// Helper for unit tests.
+        /// TODO: refactor this after adopting emitter/parser pattern.
+        /// </summary>
+        public Dictionary<string, Type>? SerializableTypes => _helper.GetSerializableTypes();
+
+        /// <summary>
+        /// Registers a syntax resolver to receive compilation units.
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new JsonSerializableSyntaxReceiver());
+        }
+
+        /// <summary>
+        /// Generates source code to optimize serialization and deserialization with JsonSerializer.
+        /// </summary>
+        /// <param name="executionContext"></param>
+        public void Execute(GeneratorExecutionContext executionContext)
+        {
+            Compilation compilation = executionContext.Compilation;
+
+            const string JsonSerializableAttributeName = "System.Text.Json.Serialization.JsonSerializableAttribute";
+            INamedTypeSymbol jsonSerializableAttribute = compilation.GetTypeByMetadataName(JsonSerializableAttributeName);
```

Comment:
What happens if one of these was also defined in the dev's project in addition to in the core library... does that impact this?

Reply:
Following the latest change - if the custom attribute has the exact same shape as ours **&** our attribute is also in the compilation, then the generator will honor it. Otherwise, the generator will ignore it. Added couple tests.